File: app/src/types/ai-create.ts
// The Create-with-AI conversation contract. Mirrored by ConverseIn/ConverseOut in
// veodyn_api/schemas/ai.py and by the Zod schema in
// src/app/api/ai/converse/route.ts, so a change here has to land in all three.
// docs/docs/operations/ai-provider.md carries the contract.
//
// Unlike /ai/outline, the request carries NO grounding: the service assembles the
// catalog and the query list itself.
//
// `CreateKind` names all five kinds because the SERVICE contract does
// (schemas/ai_create.py is community), but `kpi` and `report` are owned by
// installed features: their wire shapes live beside the code that renders them
// (`KpiProposal` in @/types/kpi, `ReportProposal` in @/types/ai-report) and reach
// the chat through `FeatureDescriptor.proposals`.

export type CreateKind = 'query' | 'dashboard' | 'kpi' | 'report' | 'snippet'

export interface ConverseMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface ConverseRequest {
  kind: CreateKind
  messages: ConverseMessage[]
  // The dashboard being EDITED rather than created. An id and nothing else: the
  // service reads what the dashboard actually holds.
  targetDashboardId?: number
  // The table the previous turn said it profiled, handed back so the next turn
  // profiles the same one. The endpoint is stateless, so this round trip is the
  // whole mechanism. A name only, and the service checks it against the catalog.
  focusTable?: string | null
}

export interface QueryProposal {
  kind: 'query'
  name: string
  description: string
  sql: string
  datasetTable: string
  // A VIZ_CHOICES id (src/lib/viz-choices.ts), not a Redash type: a chart shape
  // is an *option* on a CHART visualization. resolveVizChoice turns this into
  // {type, options} and falls back to the table for an id this build no longer
  // offers.
  vizChoiceId: string
  // Renderer options the service authored for that shape against the columns the
  // statement actually returns, already sanitized against the same per-type
  // allowlist a published report goes through.
  //
  // Always present, often empty: the wire always carries the key.
  vizOptions: Record<string, unknown>
}

/**
 * A query the AI wrote because the instance does not have one that fits.
 *
 * QueryProposal's fields minus its discriminator, and its SQL came from the same
 * generator, so it went through the same safety check. Three things carry one: a
 * dashboard widget, a KPI, and `OutlineSection.newQuery` in ai-report.ts.
 */
export interface NewQueryProposal {
  name: string
  description: string
  sql: string
  datasetTable: string
  vizChoiceId: string
  // As QueryProposal.vizOptions, and required for the same reason.
  vizOptions: Record<string, unknown>
}

/**
 * One widget: a query that exists, or one to be written first.
 *
 * Exactly one of the pair, which the service enforces before it answers. Both
 * halves are always present on the wire (the unused one null), so the card reads
 * one shape and decides by which is filled in.
 */
export interface DashboardWidgetProposal {
  title: string
  queryId: number | null
  visualizationId: number | null
  /**
   * The shape this widget should be drawn as, from `viz-choices.ts`.
   *
   * Null means "as it is drawn now", which is what an unchanged widget of an
   * edit carries. Beside a `visualizationId` it names the shape that id already
   * is, so the diff can tell a real change from a repeated widget. WITHOUT one
   * the query has no visualization of that shape and the client has to create
   * it, which is the only path here that writes to a saved query and therefore
   * the one that checks who owns it first.
   */
  vizChoiceId: string | null
  newQuery: NewQueryProposal | null
}

export interface DashboardProposal {
  kind: 'dashboard'
  name: string
  widgets: DashboardWidgetProposal[]
}

export interface SnippetProposal {
  kind: 'snippet'
  trigger: string
  snippet: string
  description: string
}

/** The kinds this build has a card for, whatever features are installed. */
export type Proposal = QueryProposal | DashboardProposal | SnippetProposal

/**
 * A proposal whose kind belongs to an installed feature.
 *
 * Opaque here: the discriminator is all the community half knows, because the
 * fields under it are the feature's contract with its own card.
 * `FeatureDescriptor.proposals` is where a feature claims a kind and supplies
 * the card; `src/features/proposals.ts` routes to it.
 *
 * The index signature keeps this from absorbing the three kinds above: an
 * interface is not assignable to an index-signature type, so `isCommunityProposal`
 * (proposal-model.ts) stays the only way across.
 */
export interface FeatureProposal {
  kind: string
  readonly [field: string]: unknown
}

/**
 * Anything a converse turn may carry: a kind this build renders itself, or one an
 * installed feature does. Held loosely from the relay to ProposalPanel, which is
 * the single place that decides which it is.
 */
export type AnyProposal = Proposal | FeatureProposal

export interface ConverseResponse {
  reply: string
  suggestedAnswers: string[]
  // Invariant, enforced at both ends: ready === (proposal !== null).
  ready: boolean
  proposal: AnyProposal | null
  // Which table this turn profiled, for the client to send back on the next one.
  // Null when the conversation has not settled on one yet, never absent: the
  // relay's response schema is strict about exactly this key set.
  focusTable: string | null
}

// Shared by the composer, the relay and the service.
export const MAX_USER_TURNS = 12
export const MAX_MESSAGE_CHARS = 4_000
export const MAX_TRANSCRIPT_MESSAGES = 25

File: api/veodyn_api/schemas/ai_create.py
"""The Create-with-AI conversation contract: one request per turn, and a
proposal on the turn it converges. The single-shot AI endpoints are schemas/ai.py.

Its authority is app/src/types/ai-create.ts, field for field; the relay's Zod
copy in app/src/app/api/ai/converse/route.ts checks the same shape on arrival
and is unforgiving, so a drift here surfaces as "AI is broken" rather than as a
tolerated difference. Change one, change the other (docs/ai-provider.md).

Imports from schemas/ai.py and never the reverse.
"""

from typing import Annotated, Any, Literal

from pydantic import Field, field_validator, model_validator

from veodyn_api.schemas.ai import CamelModel, NewQueryProposalOut, ReportOutlineOut

CreateKind = Literal["query", "dashboard", "kpi", "report", "snippet"]
# Mirrors KpiDirection/KpiCadence in app/src/types/kpi.ts. Named aliases so the
# service can clamp a model's pick to one of them without a cast.
KpiDirection = Literal["higher-is-better", "lower-is-better"]
KpiCadence = Literal["hourly", "daily", "weekly"]

# Spec section 4.3, the same constants as ai-create.ts. Enforced here as well as
# in the composer and the relay, which are the two an attacker can replace.
MAX_USER_TURNS = 12
MAX_MESSAGE_CHARS = 4_000
MAX_TRANSCRIPT_MESSAGES = 25
MAX_SUGGESTED_ANSWERS = 4


class ConverseMessageIn(CamelModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=MAX_MESSAGE_CHARS)


class ConverseIn(CamelModel):
    kind: CreateKind
    messages: list[ConverseMessageIn] = Field(min_length=1, max_length=MAX_TRANSCRIPT_MESSAGES)
    # The dashboard being EDITED. An id and nothing else: the widgets it holds
    # are read from Redash as the caller, never accepted from the browser.
    target_dashboard_id: int | None = Field(default=None, gt=0)
    # The table the conversation has settled on, handed back from the previous
    # turn. Only a LOOKUP KEY, resolved against the catalog this service
    # assembles itself, so a forged value can at worst match nothing and cost the
    # turn its profile.
    focus_table: str | None = Field(default=None, max_length=255)

    @field_validator("messages")
    @classmethod
    def _within_the_turn_cap(cls, messages: list[ConverseMessageIn]) -> list[ConverseMessageIn]:
        """The user-turn cap, which counts a subset and so cannot be a Field
        constraint. Truncating the transcript instead would drop the opening
        turns, which is where the goal is."""
        turns = sum(1 for message in messages if message.role == "user")
        if turns > MAX_USER_TURNS:
            raise ValueError(f"a conversation may not exceed {MAX_USER_TURNS} user turns")
        return messages


class QueryProposalOut(CamelModel):
    kind: Literal["query"] = "query"
    name: str
    description: str
    sql: str
    dataset_table: str
    # A VIZ_CHOICES id, not a Redash type. See CHOICE_IDS in
    # services/ai_viz_choice.py for the closed list and where it comes from.
    viz_choice_id: str
    # The renderer options for that shape (a column mapping, stacking, an axis
    # scale), sanitized against schemas/public_viz_options.py and checked against
    # the statement's real columns before it gets here.
    #
    # No default: pydantic marks a defaulted field OPTIONAL in the OpenAPI
    # schema, openapi-typescript then generates `vizOptions?`, and the generated
    # contract test refuses it against the hand-written type. The wire always
    # carries this key.
    viz_options: dict[str, Any]


class DashboardWidgetProposalOut(CamelModel):
    """One widget: either a query that exists, or one to be written first.

    Exactly one of the two, enforced below: a widget carrying both would let the
    card create a query and then hang a different one on the dashboard, and one
    carrying neither is a row with nothing behind it.
    """

    # No defaults: all five are present on the wire with the unused half null. A
    # default would make them optional in the generated schema, and the client
    # would have to read two shapes (absent, or null) for one fact.
    title: str
    query_id: int | None
    visualization_id: int | None
    # The shape this widget should be drawn as, from ai_viz_choice.CHOICE_IDS.
    # Null means "as it is drawn now". Set with `visualization_id` it names the
    # shape that id already is; set WITHOUT one, the client must create the
    # visualization, which is the only path here that writes to a saved query and
    # so the one that has to check who owns it first.
    viz_choice_id: str | None
    new_query: NewQueryProposalOut | None

    @model_validator(mode="after")
    def _exactly_one_source(self) -> "DashboardWidgetProposalOut":
        existing = self.query_id is not None
        if existing == (self.new_query is not None):
            raise ValueError("a widget names either an existing query, or a new one")
        if existing and self.visualization_id is None and not self.viz_choice_id:
            raise ValueError("a widget over an existing query needs a visualization to show or a shape to create")
        return self


class DashboardProposalOut(CamelModel):
    kind: Literal["dashboard"] = "dashboard"
    name: str
    widgets: list[DashboardWidgetProposalOut]


class KpiProposalOut(CamelModel):
    """A KPI over a query that exists, or over one written for it.

    A KPI reads one number from one query, so it needs exactly one source. Both
    halves are always on the wire with the unused one null, the same rule
    DashboardWidgetProposalOut follows and for the same reason.
    """

    kind: Literal["kpi"] = "kpi"
    name: str
    description: str
    source_query_id: int | None
    new_query: NewQueryProposalOut | None
    # Not checkable without running the query, which this service will not do,
    # so it is a plain string the card surfaces as an editable field.
    value_column: str
    unit: str | None
    target: float | None
    direction: KpiDirection
    cadence: KpiCadence
    # The bands, when the analyst asked for them explicitly. Null is the ordinary
    # case and means "derive from the target", which is what the card does.
    #
    # Required rather than defaulted, so they are `required` in the OpenAPI
    # schema and the generated TypeScript reads `number | null` rather than
    # optional.
    at_risk: float | None
    breached: float | None

    @model_validator(mode="after")
    def _exactly_one_source(self) -> "KpiProposalOut":
        if (self.source_query_id is not None) == (self.new_query is not None):
            raise ValueError("a KPI names either an existing query or a new one")
        return self

    @model_validator(mode="after")
    def _coherent_bands(self) -> "KpiProposalOut":
        """Drop a pair that would erase a band, rather than refusing the KPI.

        `statusForValue` reads the two in direction order, so for
        higher-is-better a breached bound ABOVE the at-risk one makes the at-risk
        band unreachable. Dropped to null (the target then derives them) rather
        than raised, so the whole proposal is not lost over a detail the card can
        fill in. Both or neither, since one bound alone has no band to end.
        """
        if self.at_risk is None or self.breached is None:
            self.at_risk = None
            self.breached = None
            return self
        worse_is_lower = self.direction == "higher-is-better"
        ordered = self.breached <= self.at_risk if worse_is_lower else self.breached >= self.at_risk
        if not ordered:
            self.at_risk = None
            self.breached = None
        return self


class ReportProposalOut(CamelModel):
    kind: Literal["report"] = "report"
    outline: ReportOutlineOut


class SnippetProposalOut(CamelModel):
    kind: Literal["snippet"] = "snippet"
    trigger: str
    snippet: str
    description: str


# The bare union is what services annotate against; the discriminated form is
# what the field uses.
AnyProposalOut = QueryProposalOut | DashboardProposalOut | KpiProposalOut | ReportProposalOut | SnippetProposalOut

# Discriminated on `kind`, so a proposal that does not match its own kind is a
# validation failure here rather than a card the frontend cannot render.
ProposalOut = Annotated[AnyProposalOut, Field(discriminator="kind")]


class ConverseOut(CamelModel):
    reply: str
    # All five required, and `proposal` required-and-nullable rather than
    # optional: the relay's schema demands the key be present, and a default here
    # would put an optional field in the OpenAPI schema that the generated
    # frontend contract then reads as "may be absent".
    suggested_answers: list[str] = Field(max_length=MAX_SUGGESTED_ANSWERS)
    ready: bool
    proposal: ProposalOut | None
    # The table this conversation is about, for the client to send back on the
    # next turn. No default, for the reason above.
    focus_table: str | None

    @model_validator(mode="after")
    def _ready_means_a_proposal(self) -> "ConverseOut":
        """There is exactly one way to be ready.

        Enforced here so a UI cannot render a Create button for nothing and a
        "here is your dashboard" reply cannot arrive empty. The relay refuses
        both directions too; this is the copy that runs before the bytes leave.
        """
        if self.ready != (self.proposal is not None):
            raise ValueError("ready must be true if and only if a proposal is present")
        return self
